### DISCUSSION TITLE
[JS101] 綜合題目練習Lv3  練習一：排序

### DISCUSSION THREAD

**Discussion by KW0409** [Lidemy/forum#105] [前端相關問題] [answered]
大家好，這題我先嘗試用下面的選擇排序法正確解出來後
```
function sort(arr){
  for(let i=0; i<arr.length; i++){
    let minIndex = i
    let min = arr[minIndex]
    for(let j=i+1; j<arr.length; j++){
      if(arr[j] < min){
        min = arr[j]
        minIndex = j
      }
    }
    [arr[minIndex],arr[i]] = [arr[i],arr[minIndex]]
  }
  return arr
}
console.log(sort([6, 8, 3, 2]))
```
試著把 `[arr[minIndex],arr[i]] = [arr[i],arr[minIndex]]` 的位置改變
也就是變成下面的樣子
```
function sort(arr){
  for(let i=0; i<arr.length; i++){
    let minIndex = i
    let min = arr[minIndex]
    for(let j=i+1; j<arr.length; j++){
      if(arr[j] < min){
        min = arr[j]
        minIndex = j
        [arr[minIndex],arr[i]] = [arr[i],arr[minIndex]]
      }
    }
  }
  return arr
}
console.log(sort([6, 8, 3, 2]))
```
但是跑出來的結果卻會是錯的
用 chrome 的 debugger 看了發現是執行到 `[arr[minIndex],arr[i]] = [arr[i],arr[minIndex]]` 的時候交換值會出問題，
但是看不懂為什麼會出錯
麻煩各位指點迷津～

**Comment by allenliaooooooo** [comment, marked as answer]
> ```
>         minIndex = j
>         [arr[minIndex],arr[i]] = [arr[i],arr[minIndex]]
> ```

在 `minIndex = j` 後面加上 `;` 就會正確了
我猜是這兩行被解讀成同一行了，變成 `minIndex = j[arr[minIndex],arr[i]] = [arr[i],arr[minIndex]]`

另外，其實 `minIndex` 跟 `min` 都可以省略不用宣告，直接用 `i` 跟 `arr[i]` 就 OK 嘍

**Reply by KW0409** [reply]
誒我記得昨天我也試過加 `;` 的方法
但是我忘了加在哪，可能加錯地方了哈哈
不過我可以再問問這兩行會被解讀成同一行的原因嗎？

**Reply by allenliaooooooo** [reply]
JavaScript 自動加 `;` 的規則我也不清楚
可能要請更了解的人說明了

**Reply by allenliaooooooo** [reply]
可以參考這個影片 https://www.youtube.com/watch?v=B4Skfqr7Dbs
5:17 開始有相關 case

**Reply by KW0409** [reply]
好的，還是非常感謝你解決我的疑惑
謝謝！

**Reply by allenliaooooooo** [reply]
多虧你的發問我才去研究，也才會學到這個哈哈
